processor/deltatocumulativeprocessor/reorderprocessor/processor.go

```diff
@@ -0,0 +1,198 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package reorderprocessor // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/reorderprocessor"
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/reorderprocessor/internal/gmetric"
+)
+
+type Config struct {
+	Hold time.Duration `mapstructure:"hold"`
+	Max  int           `mapstructure:"max_streams"`
+}
+
+func Default() Config {
+	return Config{
+		Hold: 1 * time.Minute,
+		Max:  0, // no limit
+	}
+}
+
+type Processor struct {
+	cfg  Config
+	next consumer.Metrics
+	log  *zap.Logger
+
+	ctx    context.Context
+	cancel func()
+
+	mtx  sync.Mutex
+	meta Meta
+
+	nums Queue[pmetric.NumberDataPoint, pmetric.NumberDataPointSlice]
+	expo Queue[pmetric.ExponentialHistogramDataPoint, pmetric.ExponentialHistogramDataPointSlice]
+	hist Queue[pmetric.HistogramDataPoint, pmetric.HistogramDataPointSlice]
+	summ Queue[pmetric.SummaryDataPoint, pmetric.SummaryDataPointSlice]
+}
+
+func New(cfg *Config, next consumer.Metrics, log *zap.Logger) *Processor {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	proc := Processor{
+		cfg:  *cfg,
+		next: next,
+		log:  log,
+
+		ctx:    ctx,
+		cancel: cancel,
+
+		mtx: sync.Mutex{},
+		meta: Meta{
+			resources: make(map[identity.Resource]pcommon.Resource),
+			scopes:    make(map[identity.Scope]pcommon.InstrumentationScope),
+			metrics:   make(map[identity.Metric]pmetric.Metric),
+		},
+
+		nums: make(Queue[pmetric.NumberDataPoint, pmetric.NumberDataPointSlice]),
+		hist: make(Queue[pmetric.HistogramDataPoint, pmetric.HistogramDataPointSlice]),
+		expo: make(Queue[pmetric.ExponentialHistogramDataPoint, pmetric.ExponentialHistogramDataPointSlice]),
+		summ: make(Queue[pmetric.SummaryDataPoint, pmetric.SummaryDataPointSlice]),
+	}
+
+	return &proc
+}
+
+func (p *Processor) size() int {
+	return len(p.nums) + len(p.expo) + len(p.hist) + len(p.summ)
+}
+
+func (p *Processor) ConsumeMetrics(_ context.Context, md pmetric.Metrics) error {
```

Comment:
@RichieSams I tried to go the "non-layered" route this time and think it turned out quite well. control flow is very visible from `ConsumeMetrics` (apart from `moveToQueue` being its own function, because no anonymous generic functions for us, thanks Go).

This certainly has less complexity than the deltatocumulative map overloading .. xd

wdyt?

Reply:
I like it! 😁